config-model/src/main/java/com/yahoo/vespa/model/application/validation/change/IndexingModeChangeValidator.java

```diff
@@ -27,31 +35,56 @@
         for (Map.Entry<String, ContentCluster> currentEntry : currentModel.getContentClusters().entrySet()) {
             ContentCluster nextCluster = nextModel.getContentClusters().get(currentEntry.getKey());
             if (nextCluster == null) continue;
-
-            Optional<ConfigChangeAction> change = validateContentCluster(currentEntry.getValue(), nextCluster, overrides, now);
-            if (change.isPresent())
-                actions.add(change.get());
+            actions.addAll(validateContentCluster(currentEntry.getValue(), nextCluster, overrides, now));
         }
         return actions;
     }
 
-    private Optional<ConfigChangeAction> validateContentCluster(ContentCluster currentCluster, ContentCluster nextCluster,
-                                                                ValidationOverrides overrides, Instant now) {
-        boolean currentClusterIsIndexed = currentCluster.getSearch().hasIndexedCluster();
-        boolean nextClusterIsIndexed = nextCluster.getSearch().hasIndexedCluster();
-
-        if (currentClusterIsIndexed == nextClusterIsIndexed) return Optional.empty();
+    private List<ConfigChangeAction> validateContentCluster(
+            ContentCluster currentCluster, ContentCluster nextCluster, ValidationOverrides overrides, Instant now) {
+        List<ConfigChangeAction> changes = new ArrayList<>();
+        ContentSearchCluster currentSearchCluster = currentCluster.getSearch();
+        ContentSearchCluster nextSearchCluster = nextCluster.getSearch();
+        {
+            Set<String> currentStreamingTypes = toDocumentTypeNames(currentSearchCluster.getDocumentTypesWithStreamingCluster());
+            Set<String> nextIndexedTypes = toDocumentTypeNames(nextSearchCluster.getDocumentTypesWithIndexedCluster());
+            for (String type : nextIndexedTypes) {
+                if (currentStreamingTypes.contains(type)) {
+                    changes.add(createReindexAction(overrides, now, nextCluster, type, "streaming", "indexed"));
+                }
+            }
+        }
+        {
+            Set<String> currentIndexedTypes = toDocumentTypeNames(currentSearchCluster.getDocumentTypesWithIndexedCluster());
+            Set<String> nextStreamingTypes = toDocumentTypeNames(nextSearchCluster.getDocumentTypesWithStreamingCluster());
+            for (String type : nextStreamingTypes) {
+                if (currentIndexedTypes.contains(type)) {
+                    changes.add(createReindexAction(overrides, now, nextCluster, type, "indexed", "streaming"));
+                }
+            }
+        }
+        return changes;
```

Comment:
We'll miss changes from "index" to "store-only", or vv.

Reply:
Ok, I can add that as well

Reply:
Ok, later.